### ISSUE TITLE
Nested variable visibility in complex rule conditions

### ISSUE THREAD

**Issue by mrrodriguez** [rbrush/clara-rules#97]
This is related to the changes in https://github.com/rbrush/clara-rules/issues/90.
The case here is involves variable visibility inside of "complex conditions" in rules, such as `:not`, `:and`, and `:or`.

I think these issues are specific to variable bindings in nested expressions that are automatically moved around for non-equals (unifying) "pseudo - join bindings".  It looks like the Clara compiler will automatically move these non-equality binding constraints into a auto-generated `:test` node.

---

I'd expect these few cases to work:

`:or` node condition:

```clj
(let [any-cold? (dsl/parse-rule [[Temperature (= ?t temperature)]
                                 [:or
                                  [Cold (< temperature ?t)]
                                  [Cold (< temperature 5)]]]

                                (insert! ^{:type :found-cold} {:found true}))

      found-cold (dsl/parse-query [] [[:found-cold]])

      session (-> (mk-session [any-cold? found-cold])
                  (insert (->Temperature 10 "MCI")
                          (->Cold 5))
                  fire-rules)

      results (query session found-cold)]
  
  (is (= 1 (count results)))

  (is (= ^{:type found-cold} {:found true}
         (first results))))


;; Returns
ExceptionInfo Using variable that is not previously bound. This can happen when a test expression uses a previously unbound variable, or if a a variable is referenced in a nested part of a parent expression, such as (or (= ?my-expression my-field) ...). Unbound variables: #{?t}  clojure.core/ex-info (core.clj:4403)
```

Side note: the small typo in that error message "...or if a a variable is..." - has two "a"s.

--

Also with a `:not` node condition:

```clj
(let [colder-temp (dsl/parse-rule [[Temperature (= ?t temperature)]
                                   [:not [Cold (or (< temperature ?t)
                                                   (< temperature 0))]]]
                                  
                                  (insert! ^{:type :found-colder} {:found true}))

      find-colder (dsl/parse-query [] [[:found-colder]])

      session (-> (mk-session [colder-temp find-colder])
                  (insert (->Temperature 10 "MCI")
                          (->Cold 11))
                  fire-rules)

      results (query session find-colder)]

  (is (= 1 (count results)))

  (is (= {}
         (first results))))

;; Returns
CompilerException java.lang.RuntimeException: Unable to resolve symbol: ?t in this context, compiling:(some/tmp/path.clj:2:48) 

```

--

I believe that `:and` node conditions actually are not seeing this issue, but I haven't dug into that path too deeply.

---

I haven't gotten a chance yet to look in depth into the source of the problem.

**Comment by rbrush** [comment]
Yes, it looks like the check for an unbound variable is creating some false
positives on these scenarios. I'll take a closer look.

**Comment by rbrush** [comment]
Alright, so this was due to the fact that converting the expression to disjunctive normal form changed the expression order as a side effect. I moved the test to check after conditions were properly sorted, which addresses the issue. I wasn't quite ready to merge into master for this (and need to run, life happens), but I'm going to review this further and get it merged tonight.

**Comment by rbrush** [comment]
Should be fixed in the above commit, which I went ahead and merged into master. Negated conditions were actually never supported with arbitrary tests, so I went ahead and added support, which included a specialized negation node that is analogous to what I did with accumulators. All tests pass, including several use cases exploring behavior described here.

I'm going to go ahead and close this since I'm pretty sure I got these scenarios, but we can reopen or open a separate issue if I missed something.
